api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java

```diff
@@ -0,0 +1,324 @@
+          thread.setDaemon(true);
+          thread.setName("oauth2-client");
+          return thread;
+        });
+  }
+
+  @Override
+  public AccessToken authenticate() {
+    return getCurrentTokens().getAccessToken();
+  }
+
+  /** Visible for testing. */
+  Tokens getCurrentTokens() {
+    try {
+      return currentTokensStage.toCompletableFuture().get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RuntimeException) {
+        throw (RuntimeException) cause;
+      } else if (cause instanceof Error) {
+        throw (Error) cause;
+      } else {
+        throw new RuntimeException(cause);
+      }
+    }
+  }
+
+  /**
+   * Starts the client.
+   *
+   * <p>Calling this method will trigger a first access token fetch, using the "client_credentials"
+   * grant type. It will also schedule a refresh of the access token when needed.
+   *
+   * <p>Calling this method twice or more is a no-op.
+   */
+  public void start() {
+    started.complete(null);
+  }
+
+  @Override
+  public void close() {
+    try {
+      currentTokensStage.toCompletableFuture().cancel(true);
+      ScheduledFuture<?> tokenRefreshFuture = this.tokenRefreshFuture;
+      if (tokenRefreshFuture != null) {
+        tokenRefreshFuture.cancel(true);
+      }
+      if (shouldCloseExecutor) {
+        if (!executor.isShutdown()) {
+          executor.shutdown();
+          if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+            executor.shutdownNow();
+          }
+        }
+      }
+      if (password != null) {
+        Arrays.fill(password, (byte) 0);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  private void scheduleTokensRenewal(Tokens currentTokens) {
+    Instant now = Instant.now();
+    Instant accessExpirationTime =
+        tokenExpirationTime(now, currentTokens.getAccessToken(), defaultAccessTokenLifespan);
+    Instant refreshExpirationTime =
+        tokenExpirationTime(now, currentTokens.getRefreshToken(), defaultRefreshTokenLifespan);
+    Duration delay =
+        nextDelay(now, accessExpirationTime, refreshExpirationTime, refreshSafetyWindow);
+    LOGGER.debug("Scheduling token refresh in {}", delay);
+    tokenRefreshFuture =
+        executor.schedule(
+            () -> {
+              currentTokensStage = renewTokens();
+              currentTokensStage.thenAccept(this::scheduleTokensRenewal);
+            },
+            delay.toMillis(),
+            TimeUnit.MILLISECONDS);
+  }
+
+  private CompletionStage<Tokens> renewTokens() {
+    return currentTokensStage
+        // try refreshing the current tokens
+        .thenApply(this::refreshTokens)
+        // if that fails, try fetching brand-new tokens
+        .exceptionally(error -> fetchNewTokens());
```

Comment:
Does it make sense to log the error here or any (runtime) exception in `refreshTokens`?
Guess, it can only be either the HTTP request failing or the response not being deserializable in `refreshTokens` (and `exchangeTokens`).